### PR TITLE
fix rsync for paths with spaces

### PR DESCRIPTION
**Pull request by infwinston**
This PR fixes https://github.com/skypilot-org/skypilot/issues/1189.

Test:
- [x] The example in the issue

### CODE REVIEW

**Review by concretevitamin — COMMENTED** [review]
Does `{source!r}` etc work? which is supposedly more smart? Please also add a tested: section in the PR.

**Comment by infwinston** [comment]
> Does `{source!r}` etc work? which is supposedly more smart? Please also add a tested: section in the PR.

Good idea let's use `!r`. I only tested with a simple example with space but not sure if it's enough.

**Comment by infwinston** [comment]
After some careful tests, I found that with the quotes, `~` would not work because it will not be correctly interpreted by the shell. alternative way is to replace all the spaces with `\ `. Not sure whether this covers all cases..

**Comment by Michaelvll** [comment]
> After some careful testing, I found that with the quotes, path like `~/tmpfile: ~/tmpfile` would not work because `~` will not be correctly interpreted by the shell in the quotes.
> 
> Alternative way is to replace all the spaces with `\ ` and remove quotes. But still not sure whether this covers all cases..

Another way is to use `"` instead of `!r`, as the `!r` can add `'` as the quote, which has different effect in shell against `"`.

**Comment by concretevitamin** [comment]
How about `os.path.expanduser()` first then do `!r`?  Note that on Mac, files `"` and `''` are both valid.

**Comment by infwinston** [comment]
Yeah `'` and `"` have the same issue I think. Both of them won't be correctly interpreted by rsync.

I tested `os.path.expanduser()`  and there is still one issue. 
Basically one of our smoke tests (`test_file_mounts`) won't pass.
The reason is due to the change of semantic for path `.` by `expanduser()`.
```
file_mounts:
  /sky: .
```

Without `expanduser()` we basically upload all files to `/sky/`.
But with `expanduser()` we will upload the entire directory to `/sky/` so it becomes `/sky/THE-DIRECTORY/...`

I'm not entirely sure if we want to change the semantic of `.`?

**Comment by concretevitamin** [comment]
cc @Michaelvll

**Comment by Michaelvll** [comment]
> I tested `os.path.expanduser()` and there is still one issue. Basically one of our smoke tests (`test_file_mounts`) won't pass. https://github.com/skypilot-org/skypilot/blob/master/examples/using_file_mounts.yaml#L29 The reason is due to the change of semantic for path `.` by `expanduser()`.

Do we have any idea of why that happens, as `os.path.expanduser('.')` returns the same `'.'`?

**Comment by infwinston** [comment]
sorry for the late response. The reason is `rsync` treats `.` specially. basically the semantic of `rsync .` and `rsync $PWD` is different.
some reference: https://superuser.com/questions/905126/rsync-copying-current-directory-with-name

**Comment by infwinston** [comment]
OK I've updated the code and the issue I mentioned should be fixed now. It's solved by adding a trailing slash when needed. @Michaelvll does this look good to you?

**Comment by infwinston** [comment]
Update codes as per @Michaelvll  suggestion!

File: sky/utils/command_runner.py [hunk 1]
```diff
@@ -301,18 +301,23 @@ def rsync(
...
+            full_source_str = os.path.join(full_source_str, '')
+
         ssh_options = ' '.join(
             ssh_options_list(self.ssh_private_key, self.ssh_control_name))
         rsync_command.append(f'-e "ssh {ssh_options}"')
         if up:
             rsync_command.extend([
-                source,
-                f'{self.ssh_user}@{self.ip}:{target}',
+                f'{full_source_str!r}',
```
Comment (Michaelvll) [inline sky/utils/command_runner.py:314]: Can we add a comment for why we need the quote for the source and target?

**Review by Michaelvll — APPROVED** [review]
Thank you for fixing this @infwinston ! LGTM. Please run the smoke tests before merging it. : )

File: sky/utils/command_runner.py (hunk 1 above)
Comment (infwinston) [reply sky/utils/command_runner.py:318]: Added!

**Comment by infwinston** [comment]
Smoke tests passed. Merging this.

**Comment by concretevitamin** [comment]
It'd be great to have a smoke test for these corner cases; to ensure future changes don't regress.

**Comment by infwinston** [comment]
> It'd be great to have a smoke test for these corner cases; to ensure future changes don't regress.

I added a filename containing spaces to the smoke test `using_file_mounts.yaml`.
```
~/tmpfile2: ~/tmp-workdir/tmp file
```
Other cases are included in the same test.

**Pull request merged**